okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/RemotingOkHttpCall.java

```diff
@@ -273,50 +285,65 @@ public Void visit(QosException.Throttle exception) {
             @Override
             public Void visit(QosException.RetryOther exception) {
                 if (maxNumRelocations <= 0) {
-                    callback.onFailure(call, new IOException("Exceeded the maximum number of allowed redirects for "
-                            + "initial URL: " + call.request().url()));
-                } else {
-                    // Redirect to the URL specified by the exception.
-                    Optional<HttpUrl> redirectTo = urls.redirectTo(request().url(),
-                            exception.getRedirectTo().toString());
-                    if (!redirectTo.isPresent()) {
-                        callback.onFailure(call, new IOException("Failed to determine valid redirect URL for '"
-                                + exception.getRedirectTo() + "' and base URLs " + urls.getBaseUrls()));
-                    } else {
-                        Request redirectedRequest = request().newBuilder()
-                                .url(redirectTo.get())
-                                .build();
-                        client.newCallWithMutableState(redirectedRequest, backoffStrategy, maxNumRelocations - 1)
-                                .enqueue(callback);
-                    }
+                    callback.onFailure(call, new SafeIoException(
+                            "Exceeded the maximum number of allowed redirects",
+                            exception,
+                            UnsafeArg.of("requestUrl", call.request().url().toString())));
+                    return null;
+                }
+
+                // Redirect to the URL specified by the exception.
+                Optional<HttpUrl> redirectTo = urls.redirectTo(request().url(), exception.getRedirectTo().toString());
+                if (!redirectTo.isPresent()) {
+                    callback.onFailure(call, new SafeIoException(
+                            "Failed to determine valid redirect URL after receiving QoS retry other",
+                            exception,
+                            UnsafeArg.of("requestUrl", call.request().url().toString()),
+                            UnsafeArg.of("redirectToUrl", exception.getRedirectTo().toString()),
+                            UnsafeArg.of("baseUrls", urls.getBaseUrls())));
+                    return null;
                 }
+
+                log.debug("Retrying call after receiving retry other",
```

Comment:
Might as well use the precise class name here? s/retry other/QosException.RetryOther`?

Reply:
Sure, I was trying to use class name agnostic messages but happy to change if you prefer using the class name